src/Smalot/PdfParser/Page.php

```diff
@@ -272,4 +277,533 @@ public function getTextArray(Page $page = null)
 
 		return array();
 	}
+    
+	/*
+     * Gets all the text data with its internal representation of the page.
+     *
+     * @return array An array with the data and the internal representation
+     *
+     */
+    
+    public function extractRawData()
+    {  
+        $text = $this->getText();
+        /*
+         * Now you can get the complete content of the object with the text on it
+         */
+        $extractedData = [];
+        $content = $this->get("Contents");
+        if (isset($content->value)){
+            $values = $content->value;
+            $text = "";
+            foreach($values as $section){
+                $text .= $section->getContent();
+            }
+            $sectionsText = $this->getSectionsText($text); 
+            foreach ($sectionsText as $sectionText){
+                $commandsText = $this->getCommandsText($sectionText);
+                foreach ($commandsText as $command){
+                    $extractedData[] = $command;
+                }
+            }
+        } else {
+            $sectionsText = $content->getSectionsText($content->getContent());
+            foreach ($sectionsText as $sectionText){
+                $commandsText = $content->getCommandsText($sectionText);
+                foreach ($commandsText as $command){
+                    $extractedData[] = $command;
+                }
+            }
+        }
+        return $extractedData;
+    }
+    
+    /*
+     * Gets all the decoded text data with it internal representation from a page.
+     *
+     * @param array $extractedRawData the extracted data return by extractRawData or
+     *                                null if extractRawData should be called
+     *
+     * @return array An array with the data and the internal representation
+     *
+     */
+    public function extractDecodedRawData($extractedRawData = null)
+    {    
+        if (!isset($extractedRawData) or !$extractedRawData) {
+            $extractedRawData = $this->extractRawData();
+        }
+        $unicode = true;
+        $currentFont = null;
+        foreach ($extractedRawData as &$command){
```

Comment:
Using `&` with `$command` here results in a change of the entry. It would also change `$extractedRawData` to, doesn't it? Or is this only in the copy?

Reply:
Only in the Copy. Before calling the method extractDecodedRawData, you must call extractRawData.  In that method, you create a copy of the extractedData to return. This copy is the one that you are changing.

Reply:
The summary of changes (without the properties that were changed to public):
### In the Page class:
### Properties:
A protected property was added:  $dataTm. It is used to save the information used to parse for info in coordinates. It avoids to make all the process every time the user asked for a info in different coordinates in the same page.

### Methods:

Internal methods (This are important for getting the data and text matrix processed properly) 
- extractRawData: get the data in the internal representation without decode it.
- extractDecodedRawData: decoded the extractedRawData.
- getDataCommands: get all the commands that involved text positions changes or changes the Text Matrix, and the Text Matrix values itself. All other commands are discarded.

Methods that will be used directly by the user.
- getDataTM: This method, using the Text Matrix values and the text commands that involved in changing the value of the Text Matrix, have every TEXT with it proper Text Matrix Value. He returns this to the users and set the property $dataTm with this.  This method allows the user to understand WHERE is every TEXT in the page. 	The user should use this info, the first time, for every file of the same "FORMAT" or	"TYPE" to know the (X,Y) coordinate for every text in the pdf.  So After this, every File of the same "FORMAT" or "TYPE", could be parse for the known coordinates	for the data the users is looking for.
- getTextXY:  It returns the text (and the text matrix value) that are at the given (X, Y) coordinates using the $xError and $yError for defining a "NEAR" factor.  This is the method that the user will use to find the information he is looking for.  The $xError and $yError are important, because the information could be not exactly at the same (X, Y) coordinate but near.  So you have to define what is "Near", by giving an error for the coordinates.

Let me know if you need more info.

Reply:
For the requests:

> Besides that, the following changes need to be addressed:

>  src/Smalot/PdfParser/Element.php: instead of making value public, please make it protected again and add a getter-function to retrieve its value.

OK.  I will start this tomorrow or the day after tomorrow.
 
> &$command: see related comment

Let me know if my answer on that is OK for you.

> can you merge in the current master branch and run PHP-CS-Fixer to comply with our new coding styles? FYI: #292 

Before starting with the first point here (the getter function and change back the properties to protected), I will marge with the current master branch. 

For the PHP-CS-Fixser, I will need help on this. What I am supposed to do? (I will follow that).

I am know doing personal stuff, so it could take a while to do everything, but count on me on doing all that is needed, and thanks for your support, guidance and recommendations.